# Selección de ejercicios resueltos del Trabajo Práctico 3 de Referenciación en Geofísica (2026)

Luciano P. O. Mendoza (lmendoza@fcaglp.unlp.edu.ar)

## Notas

- Para **ejecutar** una celda hay que seleccionarla y oprimir SHIFT+ENTER.
- Para agregar una nueva celda abajo oprimir B o usar el botón + del menu superior.
- Para agregar una nueva celda arriba oprimir A.
- Para **guardar** el cuaderno oprimir SHIFT+S o usar el botón *floppy disk* del menú superior.
- Usar el menú desplegable superior para seleccionar entre los tres tipos de celdas posibles
  - *Code* para calculo (por ejemplo, como calculadora científica)
  - *Markdown* para texto con formato, ecuaciones matemáticas, imágenes, etc.
  - *Raw* para texto simple.

## Notas 2

- En este trabajo práctico emplearemos nuevas funciones y sentencias de Python:
  - bucles __infinitos__ con [while](https://www.w3schools.com/python/python_while_loops.asp)
  - nuevas funciones como [numpy.arctan2()](https://numpy.org/doc/stable/reference/generated/numpy.arctan2.html), [numpy.log10()](https://numpy.org/doc/stable/reference/generated/numpy.log10.html), [numpy.arccos()](https://numpy.org/doc/stable/reference/generated/numpy.arccos.html),[numpy.tan()](https://numpy.org/doc/stable/reference/generated/numpy.tan.html),[numpy.abs()](https://numpy.org/doc/stable/reference/generated/numpy.abs.html), etc., con un listado muy completo [aquí](https://numpy.org/doc/stable/reference/routines.math.html).
- Algunos ejercicios pueden resolverse _en modo calculadora_, empleando las celdas para calculos parciales, o directamente _programando_ todo el algoritmo o script que resuelve un ejercicio o inciso en una celda única, incluso definiendo _funciones_ sí parece conveniente.
- En varios ejercicios puede __resultar muy conveniente bosquejar__ las ecuaciones o el algoritmo que resuelve el problema en una hoja de papel, o en una celda auxiliar RAE, para luego __implementar el programa__ Python correspondiente... 

## Ejercicio 1

### Enunciado

Dada las longitudes

1. $+59^\circ 52' 16.58''$
2. $-67^\circ 43' 33.23''$
3. $+345^\circ 8' 17.67''$

cambiarlas a los formatos que se indican a continuación

1. DD,ddddddd... (con longitud de $0^\circ$ a $360^\circ$)
2. $\pm$DD,ddddddd... (con longitud de $-180^\circ$ a $+180^\circ$)
3. DD,ddddddd... E/O.


### Motivación

- La idea del ejercicio es recordar los cuidados a la hora de expresar ángulos en coordenadas sexagesimales o decimales...
- Pero agregando la idea de que no hay que **perder** información ni **crear** información ficticia $\leadsto$ ¡ojo con los decimales!
- En general, los **arcos** de longitudes se refieren a un **radio variable**, de acuerdo a cada paralelo... Aquí asumimos que estamos sobre el Ecuador, de modo que el radio que nos interesa es el del planeta (por ej., $R = 6371$ km). ¡Esto es importante para saber cuantos decimal son necesarios!
- También es importante recordar aquí que, aunque en castellano el separador decimal es una **coma**, en la mayoría de las calculadoras y en todos los lenguajes informáticos es un **punto**... ¡cuidado con esto!

### Resolución

Podemos emplear como calculadora científica las celdas de Python, **insertando** caldas adicionales con el botón **+** del menú superior de ser necesario.

In [1]:
# resolver acá!
1 + 1

2

In [2]:
# importamos numpy porque emplearemos funciones trigonométricas
import numpy as np

#### Item 1 - Caso DD,ddddddd (con longitud de $0^\circ$ a $360^\circ$)

Rápidamente podemos calcular

In [3]:
# pasar de grados sexagesimales a decimales
59 + 52/60 + 16.58/3600

59.871272222222224

es decir que $59 + \frac{52}{60} + \frac{16.58}{3600}=59.8712722222\ldots$

La cantidad de decimales dependerá de la calculadora o **software** que utilicemos...

La cuestión es, ¿quince decimales en el resultado son suficientes, o son demasiados, respecto a los dos decimales **significativos** originales?

Veamos, la circunferencia terrestre tiene $C = 2\pi\times 6371000$ metros, y también tiene $S = 360\times 60\times 60$ segundos, o sea que 1 seg **equivale** a $\frac{C}{S}$ metros. 

In [4]:
# metros en un segundo de arco
2*np.pi*6371000/(360*60*60) # m

30.887479623488534

es decir $\frac{C}{S}= 30.887...$ metros.

Entonces, dos decimales **significativos** en los segundos **equivalen** a  $30.887 \times 10^{-2} = 0.30887$ m.

In [5]:
# metros en 0.01 segundos de arco
0.01*30.88747962348853 # m

0.30887479623488534

Ahora, la circunferencia terrestre tiene $C = 2\pi\times 6371000$ metros, y también tiene $D = 360$ grados, o sea que 1 grado **equivale** a $\frac{C}{D}$.

In [6]:
# metros en un grado
2*np.pi*6371000/360 # m

111194.92664455873

es decir $\frac{C}{D}=111194.926$... metros ($\approx 111$ km).

Luego, esta cantidad de metros **significativos** equivalen a $\frac{0.30887}{111194.926} = 2.7777\times 10^{-6}$ decimales de grado significativos...

In [7]:
3.088747962348853e-01/111194.9266445587

2.777777777777778e-06

¿Y qué indica entonces este $2.7777\times 10^{-6}$? La clave es el exponente $-6$. Indica que seis decimales de grado **deberían** ser equivalentes a dos decimales de segundo de arco.

En efecto, podemos ver que $111194.926\times 10^{-6} = 0.11194 < 0.30887$, por lo que nos **sobran** decimales.

Probando con cinco vemos que $111194.926\times 10^{-5} = 1.1194 > 0.30887$, por lo que ahora nos **faltarían**...

Es decir, seis ($\times 10^{-6}$) decimales de grado son **equivalentes** a dos ($\times 10^{-2}$) decimales de segundo.

Entonces la respuesta **correcta** (la que **conserva** la información) es
$$
59^\circ 52' 16.58'' \leadsto 59.871272^\circ
$$
independientemente de qué calculadora o **software** utilicemos.

#### Item 1 - Caso $\pm$ DD,ddddddd... (con longitud de $-180^\circ$ a $+180^\circ$)

El análisis de los decimales **significativos** es exactamente como el anterior...

Además, como $0^\circ\le 59.871272^\circ < 180^\circ$, solo hay que agregar un signo $+$.

Entonces, la respuesta correcta es
$$
59^\circ 52' 16.58'' \leadsto +59.871272^\circ.
$$

#### Item 1 - Caso DD,ddddddd... E/O

El análisis de los decimales **significativos** es exactamente como el anterior...

Aquí E/O significan **al Este/al Oeste** del Meridiano de Greenwich.

En inglés (y en muchos **software**) se escribe E/W, es decir, East/West.

Como $0^\circ\le 59.871272^\circ < 180^\circ$ $\leadsto$ al Este...

Entonces, la respuesta correcta es
$$
59^\circ 52' 16.58'' \leadsto 59.871272^\circ\,\textrm{E}.
$$

#### Item 2 - Caso DD,ddddddd (con longitud de $0^\circ$ a $360^\circ$)

Rápidamente podemos calcular $-67.7258972222\ldots=-67 - \frac{43}{60} - \frac{33.23}{3600}$.

¡Pero hay que tener cuidado con los signos $-$, sobre todo si se está **programando**!

Es decir, el $-$ en los grados también se aplica a los minutos y los segundos sexagesimales.

In [8]:
-(67 + 43/60 + 33.23/3600)

-67.72589722222223

El análisis de los decimales **significativos** es exactamente como el item anterior...

Pero $-67.725897^\circ < 0^\circ$, por lo que hay que sumarle $360^\circ$, es decir $-67.725897^\circ + 360^\circ = 292.274103^\circ$.

Entonces la respuesta correcta es
$$
-67^\circ 43' 33.23'' \leadsto 292.274103^\circ
$$

#### Item 2 - Caso $\pm$ DD,ddddddd... (con longitud de $-180^\circ$ a $180^\circ$)

Es todo igual que en el item anterior...

Pero ahora, como $180^\circ \le -67.725897^\circ < 0^\circ$, no hay que hacer nada especial.

Entonces la respuesta correcta es
$$
-67^\circ 43' 33.23'' \leadsto -67.725897^\circ
$$

#### Item 2 - Caso DD,ddddddd... E/O

Es todo igual que en el item anterior...

Como $180^\circ \le -67.725897^\circ < 0^\circ$ $\leadsto$ al Oeste.

Entonces la respuesta correcta es
$$
-67^\circ 43' 33.23'' \leadsto 67.725897^\circ\,\textrm{O}.
$$
Importante: ¡sin el signo $-$!

## Ejercicio 7

### Enunciado

Calcule la aproximación esférica de la distancia entre las ciudades de Buenos Aires (BA) y Río de Janeiro (RJ), sabiendo que sus respectivas coordenadas **geográficas** son

$$
\left\{
\begin{aligned}
\lambda_{\textrm{BA}} &= 59^\circ\, 52'\, 16.58''\,\textrm{W} \\
\varphi'_{\textrm{BA}} &= 36^\circ\, 37' 59.49''\,\textrm{S}
\end{aligned}
\right.
$$

$$
\left\{
\begin{aligned}
\lambda_{\textrm{RJ}} &= 46^\circ\, 38'\, 19.74''\,\textrm{W} \\
\varphi'_{\textrm{RJ}} &= 23^\circ\, 32'\, 56.19''\,\textrm{S}
\end{aligned}
\right.
$$

Asumir que el radio de la Tierra es $R=6371$ km.

### Motivación

- La idea del ejercicio es intentar utilizar alguna de las herramientas de álgebra, geometría analítica o análisis matemático ya conocidas.
- También podrían utilizarse **fórmulas** (como la de [Haversine](https://en.wikipedia.org/wiki/Haversine_formula)), pero estás pueden resultar más complejas de calcular (con muchas funciones trigonométricas que evaluar, etc.) y son aproximadas.
- En este caso, la idea es simplemente aplicar la definición de **producto punto**...
- Con este método solo se necesita calcular **una** función trigonométrica (inversa), el resto del cálculo implica únicamente sumas, productos y cocientes.

### Resolución

Utilice como calculadora científica las siguientes celdas de cálculo, **insertando** caldas adicionales con el botón **+** del menú superior de ser necesario.

In [9]:
# resolver acá!
1 + 1

2

Del álgebra sabemos que el producto punto $f$ entre dos vectores $\mathbf{u},\mathbf{v}$ de $\mathbb{R}^3$ es 
$$
f=\mathbf{u}\cdot\mathbf{v} = u_1v_1+u_2v_2+u_3v_3 = |\mathbf{u}||\mathbf{v}|\cos\,\theta
$$
donde $\theta$ es el ángulo **subtendido** entre los dos vectores.

Luego, asumiendo un radio terrestre $R=6371$ km, y asumiendo que las alturas de ambas ciudades son nulas (es decir, $h'_{\textrm{BA}}=h'_{\textrm{RJ}}=0 $), debemos calcular los **vectores** $\mathbf{u}_{\textrm{BA}}$ y  $\mathbf{v}_{\textrm{RJ}}$.

Primero pasamos las longitudes y las latitudes a radianes (ojo con los signos, W y S implican -).

Para Buenos Aires es

In [10]:
longBA = -(59 + 52/60 + 16.58/3600)*np.pi/180 # rad
print(longBA)

-1.0449508276355999


In [11]:
latBA = -(36 + 37/60 + 59.49/3600)*np.pi/180 # rad
print(latBA)

-0.6393698100974825


Para Río de Janeiro es

In [12]:
longRJ = -(46 + 38/60 + 19.74/3600)*np.pi/180 # rad
print(longRJ)

-0.8140009100673399


In [13]:
latRJ = -(23 + 32/60 + 56.19/3600)*np.pi/180 # rad
print(latRJ)

-0.41100656744341435


Ahora sí podemos calcular las coordenadas cartesianas.

Para Buenos Aires es

In [14]:
xBA = (6371000 + 0)*np.cos(-1.044950827635600)*np.cos(-6.393698100974825e-01)
print(xBA)

2566215.5078268484


In [15]:
yBA =  (6371000 + 0)*np.sin(-1.044950827635600)*np.cos(-6.393698100974825e-01)
print(yBA)

-4421842.694531883


In [16]:
zBA =  (6371000 + 0)*np.sin(-6.393698100974825e-01)
print(zBA)

-3801511.035391586


Para Río de Janeiro es

In [17]:
xRJ = (6371000 + 0)*np.cos(-8.140009100673399e-01)*np.cos(-4.110065674434144e-01)
print(xRJ)

4010002.179110257


In [18]:
yRJ = (6371000 + 0)*np.sin(-8.140009100673399e-01)*np.cos(-4.110065674434144e-01)
print(yRJ)

-4246217.149707401


In [19]:
zRJ = (6371000 + 0)*np.sin(-4.110065674434144e-01)
print(zRJ)

-2545420.0912740794


Resultando entonces

$$
\mathbf{u}_{\textrm{BA}} =
\begin{pmatrix}
\phantom{+}2566215.5\\
-4421852.7\\
-3801511.0
\end{pmatrix}
\textrm{m}
\qquad
\mathbf{v}_{\textrm{RJ}} =
\begin{pmatrix}
\phantom{+}4010002.2\\
-4246217.1\\
-2545420.1
\end{pmatrix}
\textrm{m}
$$

Ahora podemos calcular $f=\mathbf{u}_{\textrm{BA}}\cdot\mathbf{v}_{\textrm{RJ}}$

In [20]:
f = 2566215.507826848*4010002.179110257 + (-4421842.694531883)*(-4246217.149707401) + (-3801511.035391586)*(-2545420.091274079) # m^2
print(f)

38743076627967.734


O sea que  $f=\mathbf{u}_{\textrm{BA}}\cdot\mathbf{v}_{\textrm{RJ}}=3.8743076628\times10^{13}\,\textrm{m}^2$

Además, sabemos que $|\mathbf{u}|=|\mathbf{v}|=R$, ya que las alturas son nulas (las ciudades están **sobre** las esfera de referencia).

Entonces, podemos calcular el ángulo $\theta= \arccos\left(\frac{f}{R^2}\right)$

In [21]:
theta = np.arccos(38743076627967.73/6371000**2) # rad
print(theta)

0.3027958998196326


O sea

$$
\theta = \arccos\left(\frac{f}{R^2}\right)=0.302795899819\,\textrm{rad}
$$

Ahora, recordando que la longitud $l$ de un **arco de círculo** que subtiende un ángulo $\theta$ es simplemente $l=\theta R$, obtenemos la distandia buscada

In [22]:
l = 3.027958998196326e-01*6371000 # m
print(l)

1929112.6777508794


O sea

$$
l = \theta R=0.302795899819\,\textrm{rad} \cdot 6371000\,\textrm{m} = 1929112.7\,\textrm{m}\approx 1930\,\textrm{km}
$$

### Resolución con programación en lenguaje Python

Un *programa* que resuelve este ejercicio completamente sería así

In [23]:
# queremos suficientes decimales
import numpy as np

# radio [m]
R = 6371000

# coordenadas geográficas
longBA = -(59 + 52/60 + 16.58/3600)*np.pi/180 # rad
latBA = -(36 + 37/60 + 59.49/3600)*np.pi/180 # rad
hpBA = 0 # m

longRJ = -(46 + 38/60 + 19.74/3600)*np.pi/180 # rad
latRJ = -(23 + 32/60 + 56.19/3600)*np.pi/180 # rad
hpRJ = 0 # m

# coordenadas cartesianas [m]
u = np.array([[(R + hpBA)*np.cos(longBA)*np.cos(latBA)],
              [(R + hpBA)*np.sin(longBA)*np.cos(latBA)],
              [(R + hpBA)*np.sin(latBA)]])
     
v = np.array([[(R + hpRJ)*np.cos(longRJ)*np.cos(latRJ)],
              [(R + hpRJ)*np.sin(longRJ)*np.cos(latRJ)],
              [(R + hpRJ)*np.sin(latRJ)]])
     
# ángulo [rad]
theta = np.arccos(np.sum(u*v)/R**2)

# distancia (km)
l = R*theta/1000
print(l)

1929.1126777508748


## Ejercicio 9

### Enunciado

**Convertir** las siguientes coordenadas cartesianas a coordenadas geodésicas (o sea, elipsoidales)

$$
\begin{aligned}
\begin{pmatrix}
x\\
y\\
z
\end{pmatrix}
&=
\begin{pmatrix}
\phantom{+}2780103.00\\
-4437418.94\\
-3629404.49
\end{pmatrix}
\,\textrm{m}\\
\lambda &=\,?\\
\varphi &=\,?\\
h &=\,?\\
\end{aligned}
$$

Utilice el elipsoide de referencia WGS84, con $a = 6378137.0$ m y $\frac{1}{f} = 298.257223563$.

### Pistas

- ¡Busque las expresiones en el libro, los apuntes teóricos o sus notas de clases!

### Motivación

- En principio, simplemente hay que aplicar la expresión dada en la teoría...
- Pero veremos que esta no es una expresión **directa** (otras sí lo son) sino **iterativa** (es decir, que requiere repeticiones).
- También veremos lo tedioso del cálculo, lo que indica que lo mejor sería programarlo, en una calculadora programable o una computadora personal...

### Resolución

Utilice como calculadora científica las siguientes celdas de cálculo Python, **insertando** caldas adicionales con el botón **+** del menú superior de ser necesario.

In [24]:
# resolver acá!
1 + 1

2

In [25]:
# por las dudas, importamos numpy!
import numpy as np

Lo primero es calcular los **valores auxiliares** requeridos

Empezamos por $b=a - fa$

In [26]:
b = 6378137 - 6378137/298.257223563 # m
print(b)

6356752.314245179


es decir
$$
b=a - fa=6378137.0\,\textrm{m} - \frac{1}{298.257223563}\cdot 6378137.0\,\textrm{m} = 6356752.314245179\,\textrm{m}
$$

luego $p=\sqrt{x^2 + y^2}$

In [27]:
p = np.sqrt(2780103**2 + (-4437418.94)**2) # m
print(p)

5236378.475595488


es decir
$$
p=\sqrt{x^2 + y^2}=\sqrt{(2780103.00\,\textrm{m})^2 + (-4437418.94\,\textrm{m})^2}=5236378.475595488\,\textrm{m}
$$

luego $\omega = \arctan\frac{bz}{ap}$

In [28]:
omega = np.arctan2(6356752.314245179*(-3629404.49),6378137*5236378.475595488) # rad
print(omega)

-0.6045177028216602


es decir
$$
\omega = \arctan\frac{bz}{ap}=\arctan\left(\frac{6356752.31425\,\textrm{m}\cdot -3629404.49\,\textrm{m}}{6378137\,\textrm{m}\cdot 5236378.4756\,\textrm{m}}\right)=-6.045177028216602\times 10^{-1}\,\textrm{rad}
$$

luego $c = \frac{a^2-b^2}{\sqrt{a^2p^2+b^2z^2}}$

In [29]:
c = (6378137**2 - 6356752.314245179**2)/np.sqrt(6378137**2*5236378.475595488**2 + 6356752.314245179**2*(-3629404.49)**2)
print(c)

0.006708956188545324


es decir
$$
c = \frac{a^2-b^2}{\sqrt{a^2p^2+b^2z^2}}=6.708956188545324\times 10^{-3}
$$

luego $\beta_1 =\arctan\frac{az}{bp}$

In [30]:
beta1 = np.arctan2(6378137*(-3629404.49),6356752.314245179*5236378.475595488) # rad
print(beta1)

-0.6076624874170519


es decir
$$
\beta_1 =\arctan\frac{az}{bp}=-6.076624874170519\times 10^{-1}\,\textrm{rad}
$$

Ahora debemos **iterar** (repetir) el algoritmo hasta que se cumpla la condición...

La iteración comienza con la asignación $\beta_0 = \beta_1$

In [31]:
beta0 = -6.076624874170519e-01
print(beta0)

-0.6076624874170519


es decir
$$
\beta_0 = \beta_1 = -6.076624874170519\times 10^{-1}\,\textrm{rad}
$$

luego con $g = 2\sin\left(\beta_0-\omega\right)-c\sin2\beta_0$

In [32]:
g = 2*np.sin(-6.076624874170519e-01 - (-6.045177028216602e-01)) - 6.708956188545324e-03*np.sin(2*-6.076624874170519e-01)
print(g)

-2.948669762525835e-08


es decir
$$
g = 2\sin\left(\beta_0-\omega\right)-c\sin2\beta_0=-2.948669762525835\times 10^{-8}
$$

luego con $g'= 2\left(\cos\left(\beta_0-\omega\right)-c\cos2\beta_0\right)$

In [33]:
gp = 2*(np.cos(-6.076624874170519e-01 - (-6.045177028216602e-01)) - 6.708956188545324e-03*np.cos(2*(-6.076624874170519e-01)))
print(gp)

1.9953202435916346


es decir
$$
g'= 2\left(\cos\left(\beta_0-\omega\right)-c\cos2\beta_0\right)=1.995320243591635
$$

y finalmente con $\beta_1 =  \beta_0 - \frac{g}{g'}$

In [34]:
beta1 = -6.076624874170519e-01 - (-2.948669762525835e-08)/1.995320243591635
print(beta1)

-0.6076624726391245


es decir
$$
\beta_1 =  \beta_0 - \frac{g}{g'} = -6.076624726391245\times 10^{-1}
$$

Ahora comprobamos la condición de **terminación** $\left| \beta_1 -  \beta_0 \right| \le 10^{-9}$

In [35]:
np.abs(-6.076624726391245e-01 - (-6.076624874170519e-01))

1.4777927326292684e-08

In [36]:
np.abs(-6.076624726391245e-01 - (-6.076624874170519e-01)) <= 1e-9 # ans = 0 es falso, ans = 1 es Verdadero

False

es decir
$$
\begin{aligned}
\left| \beta_1 -  \beta_0 \right| &=\\
\left| -6.076624726391245\times 10^{-1}  -\left(-6.076624874170519\times 10^{-1}\right) \right| &= 1.477792732629268\times 10^{-8}> 10^{-9}
\end{aligned}
$$
Debemos repetir...

La iteración comienza con la asignación $\beta_0 = \beta_1$

In [37]:
beta0 = -6.076624726391245e-01 # rad
print(beta0)

-0.6076624726391245


es decir
$$
\beta_0 = \beta_1 = -6.076624726391245\times 10^{-1}\,\textrm{rad}
$$

luego con $g = 2\sin\left(\beta_0-\omega\right)-c\sin2\beta_0$

In [38]:
g = 2*np.sin(-6.076624726391245e-01 - (-6.045177028216602e-01)) - 6.708956188545324e-03*np.sin(2*-6.076624726391245e-01)
print(g)

-7.546047120499111e-17


es decir
$$
g = 2\sin\left(\beta_0-\omega\right)-c\sin2\beta_0=-7.546047120499111\times 10^{-17}
$$

luego con $g'= 2\left(\cos\left(\beta_0-\omega\right)-c\cos2\beta_0\right)$

In [39]:
gp = 2*(np.cos(-6.076624726391245e-01 - (-6.045177028216602e-01)) - 6.708956188545324e-03*np.cos(2*(-6.076624726391245e-01)))
print(gp)

1.9953202433127961


es decir
$$
g'= 2\left(\cos\left(\beta_0-\omega\right)-c\cos2\beta_0\right)=1.995320243312796
$$

y finalmente con $\beta_1 =  \beta_0 - \frac{g}{g'}$

In [40]:
beta1 = -6.076624726391245e-01 - (-7.546047120499111e-17)/1.995320243312796
print(beta1)

-0.6076624726391245


es decir
$$
\beta_1 =  \beta_0 - \frac{g}{g'} = -6.076624726391245\times 10^{-1}
$$

Ahora comprobamos la condición de **terminación** $\left| \beta_1 -  \beta_0 \right| \le 10^{-9}$

In [41]:
np.abs(-6.076624726391245e-01 - (-6.076624726391245e-01))

0.0

In [42]:
np.abs(-6.076624726391245e-01 - (-6.076624726391245e-01)) <= 1e-9 # ans = 0 es falso, ans = 1 es Verdadero

True

es decir
$$
\begin{aligned}
\left| \beta_1 -  \beta_0 \right| &=\\
\left| -6.076624726391245\times 10^{-1}  -\left(-6.076624726391245\times 10^{-1}\right) \right| &= 0 \le 10^{-9}
\end{aligned}
$$
¡Ya no debemos iterar más!

Ahora podemos calcular las coordenadas geodésicas como $\lambda = \arctan\frac{y}{x}$ y como $\varphi = \arctan\frac{a\tan\beta_1}{b}$

In [43]:
longitud = np.arctan2(-4437418.94,2780103) # rad
print(longitud)

-1.01110937128956


In [44]:
latitud = np.arctan2(6378137*np.tan(-6.076624726391245e-01),6356752.314245179) # rad
print(latitud)

-0.6092376316842473


es decir
$$
\begin{aligned}
\lambda &= \arctan\frac{y}{x}=\arctan\frac{-4437418.94}{2780103.00}=-1.011109371289560\,\textrm{rad}\\
\varphi &= \arctan\frac{a\tan\beta_1}{b} = \arctan\frac{6378137 \tan \left(-6.076624726391245\times 10^{-1}\right)}{6356752.314245179}\\
&=
-6.092376316842473\times 10^{-1}\,\textrm{rad}
\end{aligned}
$$

luego podemos calcular $h = \left(p -a\cos\beta_1\right)\cos\varphi + \left(z - b\sin\beta_1\right)\sin\varphi$

In [45]:
h = (5236378.475595488 - 6378137*np.cos(-6.076624726391245e-01))*np.cos(-6.092376316842473e-01) + (-3629404.49 - 6356752.314245179*np.sin(-6.076624726391245e-01))*np.sin(-6.092376316842473e-01) # m
print(h)

29.869316736839785


es decir
$$
\begin{aligned}
h &= \left(p -a\cos\beta_1\right)\cos\varphi + \left(z - b\sin\beta_1\right)\sin\varphi\\
&=\left(5236378.475595488 - 6378137\cos\left(-6.076624874170519\times 10^{-1}\right)\right)\cos\left(-6.092376316842473\times 10^{-1}\right)+\\
&\phantom{=}
+\left(-3629404.49 - 6356752.314245179\sin\left(-6.076624874170519\times 10^{-1}\right)\right)\sin\left(-6.092376316842473\times 10^{-1}\right)\\
&=29.86931673683979
\end{aligned}
$$

Finalmente, pasando los ángulos a grados

In [46]:
longitud = -1.011109371289560*180/np.pi # deg
print(longitud)

-57.93229960101792


In [47]:
latitud = -6.092376316842473e-01*180/np.pi # deg
print(longitud)

-57.93229960101792


y teniendo en cuenta los decimales significativos de las coordenadas cartesianas originales, obtenemos la siguiente equivalencia

$$
\begin{aligned}
\begin{pmatrix}
x\\
y\\
z
\end{pmatrix}
&=
\begin{pmatrix}
\phantom{+}2780103.00\\
-4437418.94\\
-3629404.49
\end{pmatrix}
\,\textrm{m}\\
\lambda &= -57.9322996^\circ\\
\varphi &= -34.9067450^\circ\\
h &= +29.87\,\textrm{m}\\
\end{aligned}
$$

utilizando el elipsoide de referencia WGS84, con $a = 6378137.0$ m y $\frac{1}{f} = 298.257223563$.

### Resolución con programación en lenguaje Python


In [48]:
# importamos numpy...
import numpy as np

# coordenadas cartesianas
x = 2780103
y = -4437418.94
z = -3629404.49

# superficie de referencia (elipsoide)
a = 6378137.0
f = 1/298.257223563

# valores auxiliares
b = a - f*a
p = np.sqrt(x**2 + y**2)
omega = np.arctan2(b*z,a*p)
c = (a**2 - b**2)/(np.sqrt(a**2*p**2 + b**2*z**2))
beta1 = np.arctan2(a*z,b*p)

# comenzamos la iteración
while True:
    beta0 = beta1
    g = 2*np.sin(beta0 - omega) - c*np.sin(2*beta0)
    gp = 2*(np.cos(beta0 - omega) - c*np.cos(2*beta0))
    beta1 = beta0 - g/gp
    if np.abs(beta1 - beta0) < 1e-9:
        break # salir del loop infinito!

# coordenadas (rad)
longitud = np.arctan2(y,x)
latitud = np.arctan2(a*np.tan(beta1),b)

# coordenadas (deg)
longitudG = longitud*180/np.pi
latitudG = latitud*180/np.pi

# altura (m)
h = (p - a*np.cos(beta1))*np.cos(latitud) + (z - b*np.sin(beta1))*np.sin(latitud)

print(longitudG,latitudG,h)

-57.93229960101792 -34.90674501605309 29.869316736839785


## Ejercicio 10

### Enunciado

Si las coordenadas cartesianas de un punto tienen $n_d$ decimales significativos
1. ¿Cuantos decimales significativos $n_g$ tendrán las correspondientes
longitud $\lambda$ y latitud $\varphi'$ geográficas expresadas en grados decimales ($\pm$ DD.ddddddd)?
2. ¿Y la altura $h'$?

### Pistas

- ¡Piense en el método que utilizó Eratóstenes para calcular la circunferencia del planeta sin salir de Egipto!

### Motivación

- Acá debemos aplicar la misma idea que [Eratóstenes](https://en.wikipedia.org/wiki/Eratosthenes),
que es la misma idea que aplicamos en el Ejercicio 7, y que a su vez también aplicamos en el Ejercicio 1.
- Es decir, suponiendo un planeta esférico, los ángulos y las longitudes de arco son proporcionales, siendo $R$ la constante de proporcionalidad para los meridianos,
y $R\cos\varphi'$ para los paralelos...

### Resolución

Utilice como calculadora científica las siguientes celdas, **insertando** caldas adicionales con el botón **+** del menú superior de ser necesario.

In [49]:
# resolver acá!
1 + 1

2

In [50]:
# importamos numpy...
import numpy as np

#### Para las **latitudes** $\varphi'$:

podemos siempre expresar la frase **$n$ decimales significativos** de metro como **conocer hasta $10^{-n}$ de metro**.

Así, un decimal significativo equivale a $10^{-1}$, dos decimales a $10^{-2}$, etc...

El enunciado indica que conocemos $n_d$ decimales significativos de metro (o sea, $10^{-n_d}$), y queremos saber a cuantos $n_g$ decimales de grado (o sea, $10^{-n_g}$) equivalen (en longitud)...

Entonces aplicamos la idea de Eratóstenes, sabiendo que 360 grados equivalen a una circunferencia terrestre, es decir, a  $2\pi R$ metros, con $R = 6371$ km, por ejemplo.

¡Pero esto es solo válido para las **latitudes** $\varphi'$!

Entonces, podemos construir una simple regla de tres
$$
\begin{aligned}
2\pi\cdot 6371000\,\textrm{m} &\qquad 360\,\textrm{deg}\\
10^{-n_d}\,\textrm{m} &\qquad 10^{-n_g}\,\textrm{deg}
\end{aligned}
$$

que resulta en

$$
\begin{aligned}
10^{-n_g} &= \frac{360\times10^{-n_d}}{2\pi\cdot 6371000}\\
\log 10^{-n_g} &= \log\left(\frac{360\times10^{-n_d}}{2\pi\cdot 6371000}\right)\\
-n_g &= \underbrace{\log\left(\frac{360}{2\pi\cdot 6371000}\right)}_{\approx -5} - n_d\\
n_g&\approx n_d + 5
\end{aligned}
$$
Como sabemos que tanto $n_d$ como $n_g$ deben ser enteros, la expresión final que buscamos es $n_g = n_d + 5$.

Equivalentemente, resulta que $n_d = n_g - 5$.

Entonces, si conocemos las coordenadas cartesianas al milímetro (o sea, $n_d = 3$)

In [51]:
ng = -np.log10(360/(2*np.pi*6371000)) + 3 # nd = 3
print(ng)

8.046084972672869


es decir, las **latitudes** geográficas $\varphi'$ deberían tener $n_g \approx 3 + 5 = 8$ decimales.

#### Para las **longitudes** $\lambda$:

para las longitudes es **casi** igual...

¡Pero tenemos que tener en cuenta que los arcos se miden sobre paralelos, con radios menores a medida que aumenta la latitud!
$$
\begin{aligned}
2\pi\cdot 6371000\,\textrm{m}\cdot\cos\varphi'  &\qquad 360\,\textrm{deg}\\
10^{-n_d}\,\textrm{m} &\qquad 10^{-n_g}\,\textrm{deg}
\end{aligned}
$$
de donde resulta
$$
\begin{aligned}
10^{-n_g} &= \frac{360\times10^{-n_d}}{2\pi\cdot 6371000\cdot\cos\varphi'}\\
\log 10^{-n_g} &= \log\left(\frac{360\times10^{-n_d}}{2\pi\cdot 6371000\cdot\cos\varphi'}\right)\\
-n_g &= \underbrace{\log\left(\frac{360}{2\pi\cdot 6371000} \cdot \frac{1}{\cos\varphi'}\right)}_{\approx -5 -\log\left(\cos\varphi'\right)} - n_d\\
n_g&\approx n_d + 5 + \log\left(\cos\varphi'\right)
\end{aligned}
$$

Sabemos que tanto $n_d$ como $n_g$ deben ser enteros, pero la expresión final que buscamos no es entera, así que la aproximamos como $n_g \approx n_d + 5 + \log\left(\cos\varphi'\right)$.

Equivalentemente, resulta que $n_d \approx n_g - 5 - \log\left(\cos\varphi'\right)$.

Entonces, más cerca del Ecuador necesitaremos más decimales de grado, más cerca de los polos necesitaremos menos...

Por ejemplo, si conocemos las coordenadas cartesianas al milímetro ($n_d = 3$), para un punto a latitud $\varphi'=45$ deg,
las **longitudes** geográficas $\lambda$ deberían tener

In [52]:
ng = -np.log10(360/(2*np.pi*6371000)*1/np.cos(45*np.pi/180)) + 3
print(ng)

7.895569974840878


es decir, $n_g \approx 3 + 5 + \log\left(\cos\frac{\pi}{4}\right) = 3 + 5 + \log \frac{\sqrt{2}}{2} = 7.8$ decimales, es decir, $n_g = 8$.

Con $\varphi'=\pm 80$ deg resulta $n_g = 7$.

Con $\varphi'=\pm 89$ deg resulta $n_g = 6$.

#### Para las **alturas** $h'$:

este es el caso más sencillo, ya que tanto las coordenadas cartesianas como la altura son  magnitudes lineales.

Entonces, si conocemos $n_d$ decimales significativos en metros, la altura deberá tener exactamente $n_d$ decimales significativos.

¡Independientemente de la latitud!